037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,177 @@
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      mode: [running, scaling] # or just one of the two
+```
+
+Finally, the user should be able to define what are the [goals and the other rebalancing related configurations](https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaRebalance-reference) which are normally described via a `KafkaRebalance` custom resource.
+In any case, for rebalancing the cluster, an actual `KafkaRebalance` custom resource has to be automatically created by the cluster operator.
+
+### New `KafkaRebalanceTemplate` custom resource
+
+A new `KafkaRebalanceTemplate` custom resource could be a solution allowing the user to define the rebalancing configuration he wants but leaving to the operator the creation of the corresponding `KafkaRebalance` custom resource.
+This new CRD should be quite similar to the `KafkaRebalance`.
+The cluster operator has to copy over the configuration from the template to the actual rebalancing resource.
+Compared to the rejected alternative, it’s about exporting the rebalancing configuration out of the Kafka custom resource avoiding to pollute that resource and increase the size.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalanceTemplate
+metadata:
+  name: my-rebalance-template
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  # NOTE: mode and brokers fields are not available 
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The user has the flexibility to specify goals as well as all the other options like he already does today on a `KafkaRebalance` custom resource.
+The cluster operator will use that configuration for the `KafkaRebalance` but adding the `spec.mode` and `spec.brokers` accordingly.
+
+It would also contain a `spec.autoApproval` field, as described by this [proposal](https://github.com/strimzi/proposals/pull/56), to specify if auto-approval is desired or not when auto-rebalancing is enabled.
+The template to be used is referenced by a new `spec.kafka.autoRebalance.template` field and the cluster operator uses that template to create a corresponding `KafkaRebalance` custom resource.
+If the reference is missing, the cluster operator can create an “empty” `KafkaRebalance` custom resource (which means the optimization proposal will be based on Cruise Control default configuration). It will behave as the rejected alternative.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+    autoRebalance:
+      mode: [running, scaling] # or just one of the two
+      template: # reference to a KafkaRebalanceTemplate custom resource
```

Comment:
Meta point: When reading this the first time it would have been easier to either present this `template` at the same time you talk about the `mode`, or introduce the `KafkaRebalanceTemplate` resource first, and then explain how it relates to the `Kafka` resource.

Proposal points:
* Is it necessarily the case that users would always want to use the same template in both cases?
* For `running` what would that mean? would there some kind of timer? So would `periodically` be a better term? Also, any way to schedule this using the existing `maintenanceTimes` facility?

Reply:
>  or introduce the KafkaRebalanceTemplate resource first, and then explain how it relates to the Kafka resource.

isn't this the flow that we have now? I introduced the `KafkaRebalanceTemplate` first and then I talked about the `template` field in the `Kafka` CR to reference the actual template.

> Is it necessarily the case that users would always want to use the same template in both cases?

Yeah, I thought about this but was not sure if the user had really this need. Maybe yes, like skipping some hard goal when scaling but not during normal operations, or having a different replication throttle (see [here](https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaRebalanceSpec-reference) what we can set today as options for a rebalance).
At this point, `autoRebalance` would be more an array of modes having each one its own template.

```yaml
autoRebalance:
      - mode: scaling
        template: # reference to a KafkaRebalanceTemplate custom resource
      - mode: running
        template: # reference to a KafkaRebalanceTemplate custom resource
```

> For running what would that mean? would there some kind of timer? So would periodically be a better term?

Yeah it's about running a rebalancing overtime. I think there are other suggestions about this field from @tomncooper in another comment.

> Also, any way to schedule this using the existing maintenanceTimes facility?

IIRC the maintenance time feature is not about scheduling but it's about checking that you can do an operation at a specific time. For example, I want to do a Kafka upgrade but only if the window matches. Anyway you have to run the operation there is no scheduler for that.

Reply:
> or introduce the KafkaRebalanceTemplate resource first, and then explain how it relates to the Kafka resource.

I have got what you meant, just moving paragraphs around ;-)

Reply:
As already mentioned in a different comment, I would use `mode` as an list and `add-brokers`, `remove-brokers` and `full` as possible values (the same that we use today in the `KafkaRebalance` resource).